_pytest/recwarn.py

```diff
@@ -203,3 +208,15 @@ def __exit__(self, *exc_info):
                          "The list of emitted warnings is: {1}.".format(
                              self.expected_warning,
                              [each.message for each in self]))
+                elif self.match_expr is not None:
+                    for r in self:
+                        if issubclass(r.category, self.expected_warning):
+                            if compile(self.match_expr).search(str(r.message)):
```

Comment:
This is minor, but you are shadowing the builtin `compile`. I suggest changing to `re.compile` and `import re` instead.